### ISSUE TITLE
Timeouts: data structures & queries

### ISSUE THREAD

**Issue by Scooletz** [Particular/NServiceBus.Persistence.AzureStorage#46]
**TL;DR:**
Table scans in timeouts may impact performance.

**Details:**
The current approach for querying for next timeouts is following:

```
if (lastSuccessfulRead.HasValue)
{
    query = from c in timeoutDataTable.CreateQuery<TimeoutDataEntity>()
        where c.PartitionKey.CompareTo(lastSuccessfulRead.Value.ToString(partitionKeyScope)) >= 0
            && c.PartitionKey.CompareTo(now.ToString(partitionKeyScope)) <= 0
            && c.OwningTimeoutManager == endpointName
        select c;
}
else
{
    query = from c in timeoutDataTable.CreateQuery<TimeoutDataEntity>()
            where c.OwningTimeoutManager == endpointName
            select c;
}
```

[link to full TimeoutPersister.cs](https://github.com/Particular/NServiceBus.AzureStoragePersistence/blob/develop/src/NServiceBus.AzureStoragePersistence/Timeout/TimeoutLogic/TimeoutPersister.cs#L92)

The predicates are based on either scanning through partition keys, or scanning over a property. Both queries will issue full table scans which in environments with massive number of timeouts is inefficient.

**Proposal**
Provide a new design for timeouts based on creating secondary indexes based on the `endpointName` and the `TimeoutData.Time` to enable effective look ups by the query.
This is similar to the [Cassandra secondary indexes](http://stackoverflow.com/questions/29692738/how-do-secondary-indexes-work-in-cassandra) creating additional rows with the key based on an indexed value and a row containing identifiers of index rows.

The algorithm for inserting a timeout would be:
- write the secondary index entry pointing to the `TimeoutData.Id`
- write the timeout under identifier

The query for a specific period of time would access all the buckets based on the time period + `endpointName` gathering all the identifiers. In the second step, it'd query for timeouts by their primary key.

Advantages:
- no table scans (you can access do only 20.000 IOPS/account, this includes messages!)

Disadvantages:
- `select N+1`. It'll require picking timeouts one by one, but this can be done concurrently 

Looking forward to your comments: @WilliamBZA @Particular/azure-maintainers

**Comment by yvesgoeleven** [comment]
@Scooletz What do you mean with 'this can be done in batches?' How can we select individual items in batches?

**Comment by Scooletz** [comment]
@yvesgoeleven A poor choice of words. It should have been: "this can be done concurrently" without batches (I've updated the description).

**Comment by Scooletz** [comment]
I think we might close this as native delayed delivery will arrive soon in ASQ. Thoughts @Particular/azure-storage-persistence-maintainers
